python/tskit/trees.py

```diff
@@ -2198,6 +2214,9 @@ def newick(self, precision=14, root=None, node_labels=None):
                     "newick trees, one for each root."
                 )
             root = self.root
+        if not include_branch_lengths and node_labels is None:
+            # Force the python generator for simplicity, by specifying the default labels
+            node_labels = {i: str(i + 1) for i in self.leaves()}
```

Comment:
Hmm, ``leaves()`` or ``samples()``? I think it should be ``samples()``. Leaves is tricky, as we can have non-sample leaves. 

What does the C engine do? We should add a test for this, since we are doing this workaround.

Maybe it would be better to not do this, and to change the condition on the line below instead to

```
if node_labels is None and include_branch_lengths:
```

Then it's completely explicit and we don't need any comments explaining what we're doing.

Reply:
[We claim](https://tskit.readthedocs.io/en/latest/python-api.html?highlight=newick#tskit.Tree.newick) it does leaves only (regardless of whether they are samples), and that seems sensible to me. Some newick readers only bother with names for tips anyway. In particular, the docs currently say:

```
By default, leaf nodes are labelled with their numerical ID + 1, and internal nodes are not labelled.
```

Reply:
> Then it's completely explicit and we don't need any comments explaining what we're doing.

We would still need to set the node labels to the leaves only when `node_labels == None` for the python Newick generating code. But happy to move the logic around.

Reply:
Ah right, what you have sounds good then. We should add a test to verify that we get the same labels with and without ``include_branch_lengths`` in a corner case when we have non-sample leaves, though.